### ISSUE TITLE
MSSQL Code Completion Issues: Object Names and Functions

### ISSUE THREAD

**Issue by jetzerb** [dbeaver/dbeaver#10065]
#### System information: 
- Windows 10
- DBeaver 7.2.2.202010041557

#### Connection specification:
- Microsoft SQL Server 2017 (RTM) - 14.0.1000.169 (X64)
- Microsoft JDBC Driver for SQL Server (MSSQL)
- No tunnels or proxies (SSH, SOCKS, etc)

#### Describe the problem you're observing:
Code completion does not always offer suggestions when it should, and sometimes offers incorrect suggestions.
> This may also be an issue with other database engines, but I'm currently working on MSSQL and haven't tested any others.

#### Steps to reproduce, if exist:
```sql
use master;
go
drop database if exists codeCompleteTest;
go
create database codeCompleteTest;
go
use codeCompleteTest;
go
-- create objects in default schema "dbo"
create table xyzTable (id int);
go
create view xyzView as select * from xyzTable;
go
create procedure dbo.xyzProc @id int
as
	select @id;
go
create procedure dbo.xyz_Proc @id int
as
	select @id;
go
create function dbo.xyzScalarFunc(@id int)
returns int
as
begin
	return @id;
end;
go
create function dbo.xyz_Scalar_Func(@id int)
returns int
as
begin
	return @id;
end;
go
create function dbo.xyzTableValuedFunc(@id int)
returns table
as
	return select @id as idValue;
go
create function dbo.xyz_Table_Valued_Func(@id int)
returns table
as
	return select @id as idValue;
go

-- create objects in custom schema "util"
create schema util;
go

create table util.abcTable (nm varchar(15));
go
create view util.abcView as select * from util.abcTable;
go
create procedure util.abcProc @nm varchar(15)
as
	select @nm
go
create procedure util.abc_Proc @nm varchar(15)
as
	select @nm;
go
create function util.abcScalarFunc(@nm varchar(15))
returns varchar(15)
as
begin
	return @nm;
end;
go
create function util.abc_Scalar_Func(@nm varchar(15))
returns varchar(15)
as
begin
	return @nm;
end;
go
create function util.abcTableValuedFunc(@nm varchar(15))
returns table
as
	return select @nm as nameValue;
go
create function util.abc_Table_Valued_Func(@nm varchar(15))
returns table
as
	return select @nm as nameValue;
go

-- Now list all the objects we just created
select schema_name(schema_id) as SchemaName
      ,name                   as ObjectName
      ,type_desc              as ObjectType
from sys.objects where left(name,3) in ('xyz','abc')
order by 1, replace(stuff(name,1,3,''),'_',''), name
;
```

List of objects created:
|SchemaName|ObjectName|ObjectType|
|----------|----------|----------|
|dbo|xyz_Proc|SQL_STORED_PROCEDURE|
|dbo|xyzProc|SQL_STORED_PROCEDURE|
|dbo|xyz_Scalar_Func|SQL_SCALAR_FUNCTION|
|dbo|xyzScalarFunc|SQL_SCALAR_FUNCTION|
|dbo|xyzTable|USER_TABLE|
|dbo|xyz_Table_Valued_Func|SQL_INLINE_TABLE_VALUED_FUNCTION|
|dbo|xyzTableValuedFunc|SQL_INLINE_TABLE_VALUED_FUNCTION|
|dbo|xyzView|VIEW|
|util|abc_Proc|SQL_STORED_PROCEDURE|
|util|abcProc|SQL_STORED_PROCEDURE|
|util|abc_Scalar_Func|SQL_SCALAR_FUNCTION|
|util|abcScalarFunc|SQL_SCALAR_FUNCTION|
|util|abcTable|USER_TABLE|
|util|abc_Table_Valued_Func|SQL_INLINE_TABLE_VALUED_FUNCTION|
|util|abcTableValuedFunc|SQL_INLINE_TABLE_VALUED_FUNCTION|
|util|abcView|VIEW|

Test cases:
- `select * from schemaName.` neglects to suggest table valued functions:
![image](https://user-images.githubusercontent.com/36859020/96030386-8c91b100-0e21-11eb-89a5-8168cd8d17d1.png) ![image](https://user-images.githubusercontent.com/36859020/96030425-97e4dc80-0e21-11eb-8cc5-88cfe45ea372.png)
- `select * from  partialNameMatch` incorrectly suggests stored procedures, and neglects to suggest table valued functions
![image](https://user-images.githubusercontent.com/36859020/96030631-d8445a80-0e21-11eb-98bf-e5af779416c9.png)
- `select schemaName.` incorrectly suggests tables and views, and neglects to suggest scalar functions
![image](https://user-images.githubusercontent.com/36859020/96031020-67ea0900-0e22-11eb-9b5b-7a8c8a0832c4.png) ![image](https://user-images.githubusercontent.com/36859020/96030979-5a348380-0e22-11eb-95a5-5ce81b00a67c.png)
- `exec schemaName.` offers no suggestions of stored procedures to execute:
![image](https://user-images.githubusercontent.com/36859020/96031357-daf37f80-0e22-11eb-9a05-393b47049cba.png) ![image](https://user-images.githubusercontent.com/36859020/96031391-e47ce780-0e22-11eb-8f01-296b69b05d27.png)
- `exec partialName` only seems to offer suggestions when the typed name doesn't include the first character of the object name, and incorrectly suggests functions
![image](https://user-images.githubusercontent.com/36859020/96031598-31f95480-0e23-11eb-9ad3-cd38e0902b91.png) ![image](https://user-images.githubusercontent.com/36859020/96035261-658aad80-0e28-11eb-9340-79b10bd06b56.png) ![image](https://user-images.githubusercontent.com/36859020/96033605-e72d0c00-0e25-11eb-834c-fb0b65c0d3ce.png) 
- the "use global search" checkbox doesn't seem to work reliably for `exec` statements
![image](https://user-images.githubusercontent.com/36859020/96033817-396e2d00-0e26-11eb-9729-0ee2bdd4f9b4.png)
![image](https://user-images.githubusercontent.com/36859020/96033902-57d42880-0e26-11eb-8de0-07b72838225e.png)

**Comment by uslss** [comment]
thanks for the bug report

**Comment by LonwoLonwo** [comment]
possibly related to this ticket https://github.com/dbeaver/dbeaver/issues/10069

**Comment by LonwoLonwo** [comment]
Hello @jetzerb 
First of all, I want to say thank you for a well-structured ticket. Very convenient to test.

You raised a lot of problems at once, therefore, perhaps I will single out some points in separate tickets.

Let's go further point by point.

Most likely I will add auto-suggestion for all functions (not just tabular ones) after "SELECT * FROM  dbo...." for SQLServer.

At the moment, the auto-suggestion search for functions / procedures works well if you write at least part of the name after the schema name. Perhaps we will expand this case.

> exec partialName only seems to offer suggestions when the typed name doesn't include the first character of the object name, and incorrectly suggests functions

This is not a common case. Auto-suggestion does not work correctly because it considers "xyz" to be a table alias xyzTable

Global search does not work for SQLServer at the moment, I will will define it in a separate feature request

**Comment by LonwoLonwo** [comment]
List of SQL Server functions and procedures expanded for autocomplete in SQL Editor. 

If you want to see procedure or functions after SELECT keyword, enable "Show stored procedures in column list" in Code completion page.
![2020-12-16 10_42_07-Window](https://user-images.githubusercontent.com/45152336/102322527-43c8b680-3f90-11eb-8db7-48b347396b37.png)

Autocomplete for procedures/functions will work better with Content assist (Ctrl+Space) function after words EXEC/EXECUTE.

**Comment by uslss** [comment]
verified

**Comment by LonwoLonwo** [comment]
Fixed in 7.3.1
Will be available [here](https://dbeaver.io/) at 2020-12-21
